### ISSUE TITLE
Urgent. LocalRawGCSService is broken in 0.5

### ISSUE THREAD

**Issue by husayt** [GoogleCloudPlatform/appengine-gcs-client#20]
We had to rollback from 0.5 to 0.4.4 as `LocalRawGCSService` was broken in new version. Basically all local writes were failing with `security.AccessControlException` exception. This is I guess very serious, as effectively one can't use this whole library locally.

```
	Caused by: java.io.IOException: java.lang.reflect.InvocationTargetException
		at com.google.appengine.tools.cloudstorage.dev.LocalRawGcsService$BlobStorageAdapter.storeBlob(LocalRawGcsService.java:161)
		at com.google.appengine.tools.cloudstorage.dev.LocalRawGcsService.finishObjectCreation(LocalRawGcsService.java:336)
		at com.google.appengine.tools.cloudstorage.GcsOutputChannelImpl$1.call(GcsOutputChannelImpl.java:201)
		at com.google.appengine.tools.cloudstorage.GcsOutputChannelImpl$1.call(GcsOutputChannelImpl.java:198)
		at com.google.appengine.tools.cloudstorage.RetryHelper.doRetry(RetryHelper.java:108)
		... 74 more
	Caused by: java.lang.reflect.InvocationTargetException
		at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
		at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
		at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
		at java.lang.reflect.Method.invoke(Method.java:606)
		at com.google.appengine.tools.development.agent.runtime.Runtime.invoke(Runtime.java:130)
		at com.google.appengine.tools.cloudstorage.dev.LocalRawGcsService$BlobStorageAdapter.storeBlob(LocalRawGcsService.java:153)
		... 78 more
	Caused by: java.security.AccessControlException: access denied ("java.io.FilePermission" "D:\_code\myProject\target\myProj\WEB-INF\appengine-generated\encoded_gs_key:L2dzL2x4X2NlbnRyYxlL2FnZW5jeXNjcmlwdC9jaGFyZ2UtdHlwZXMuanNvbg" "write")
		at java.security.AccessControlContext.checkPermission(AccessControlContext.java:372)
		at java.security.AccessController.checkPermission(AccessController.java:559)
		at java.lang.SecurityManager.checkPermission(SecurityManager.java:549)
		at com.google.appengine.tools.development.DevAppServerFactory$CustomSecurityManager.checkPermission(DevAppServerFactory.java:429)
		at java.lang.SecurityManager.checkWrite(SecurityManager.java:979)
		at java.io.FileOutputStream.<init>(FileOutputStream.java:209)
		at java.io.FileOutputStream.<init>(FileOutputStream.java:171)
		at com.google.appengine.api.blobstore.dev.FileBlobStorage.storeBlob(FileBlobStorage.java:46)
		... 84 more
Caused by: java.io.IOException: java.lang.reflect
```
problem is caused most likely by this
```
Caused by: java.security.AccessControlException: access denied ("java.io.FilePermission" "D:\_code\myProject\target\myProj\WEB-INF\appengine-generated\encoded_gs_key:L2dzL2x4X2NlbnRyYxlL2FnZW5jeXNjcmlwdC9jaGFyZ2UtdHlwZXMuanNvbg" "write")
        at java.security.AccessControlContext.checkPermission(AccessControlContext.java:372)
```

**Comment by aozarov** [comment]
http://stackoverflow.com/questions/32236614/invocationtargetexception-storing-files-with-appengine-gcs-client-0-5-dev-appser/32236631

**Comment by husayt** [comment]
The link you provided doesn't solve this issue. It's a partial workaround. Memory backed blobstore is not the solution due to obvious limitations. 

As it stands now `LocalRawGCSService` is broken, and this issue is far from resolved unfortunately.

**Comment by aozarov** [comment]
@husayt I don't mind to change the status back and indeed the current solution is a workaround however the fix for it is not in the library but rather in the AE SDK which unfortunately is not yet public.

Do you still a value in having an open issue for it?
Until issue is resolved can you avoid writing large blobs to GCS in your integration tests?

**Comment by husayt** [comment]
Arie, this is a showstopper for us. We had to switch back to previous
version of GCS client. Yes, even if the issue is in AE as a result
LocalRawCGSService is broken and hence issue needs to stay until resolved.
Will be good to link to corresponding AE issue, if they have a ticket. I
would even advice switch back to using File Api until they resolve the
issue with new api.

As for this suggestion:
> "Until issue is resolved can you avoid writing large blobs to GCS in your
integration tests?"
Do you seriously advise us to rewrite all our integration tests, just
because local test environment is broken?




Best  regards,
Huseyn Guliyev

------------------------------------------------------------

**Comment by aozarov** [comment]
@husayt if updating your integration tests to avoid large blob is not an option then sticking to an earlier version of the client (0.4.x) as you already did would be your best option.

The files api is already deprecated and will disappear from the API.
In fact it is already disabled when using the dev appserver that is bundled with GCLOUD SDK
or when using the appengine maven plugin.

**Comment by husayt** [comment]
Yes, this is really annoying they deprecate the api without providing a working alternative. I don't even mention number of missing service methods which are not present in new API.

**Comment by aozarov** [comment]
> Yes, this is really annoying they deprecate the api without providing a working alternative. I don't even mention number of missing service methods which are not present in new API.

Can you please create separate issue[s] for the missing method[s].

**Comment by sotecyolo** [comment]
I'm having this issue as well. I noticed that although the error says it's a 'access denied' the actual reason is that Microsoft Windows does not allow ":" (colon) in file or directory names. Please don't use a colon as a logical separator since it breaks the functionality in Windows.

**Comment by aozarov** [comment]
This issue of 'access denied' should have been fixed since AE 1.9.27.

Indeed ":" would be a problem for Windows, though the code seems to be taking care of that (is
your application also using blobstore directly?)

If you are using a local dev appserver 1.9.27 and still see this this failure I would suggest you create a separate issue (as the scope of that issue is wider and should include blobstore).

**Comment by aozarov** [comment]
The security issue triggered by appengine_gcs_client using the blobstore storage rather than the files API (which was deprecated) was resolved in AE 1.9.27
